libnmstate/nm/device.py

```diff
@@ -53,25 +53,29 @@ def delete(context, dev):
         con_profile.delete()
 
 
-def modify(context, dev, connection_profile):
+def modify(context, nm_dev, connection_profile):
     """
     Modify the given connection profile on the device.
     Implemented by the reapply operation with a fallback to the
     connection profile activation.
     """
-    version_id = 0
-    flags = 0
-    action = f"Reapply device config: {dev.get_iface()}"
-    context.register_async(action)
-    user_data = context, dev, action
-    dev.reapply_async(
-        connection_profile,
-        version_id,
-        flags,
-        context.cancellable,
-        _modify_callback,
-        user_data,
-    )
+    nm_ac = nm_dev.get_active_connection()
+    if connection.is_activated(nm_ac, nm_dev):
+        version_id = 0
+        flags = 0
+        action = f"Reapply device config: {nm_dev.get_iface()}"
+        context.register_async(action)
+        user_data = context, nm_dev, action
+        nm_dev.reapply_async(
+            connection_profile,
+            version_id,
+            flags,
+            context.cancellable,
+            _modify_callback,
+            user_data,
+        )
+    else:
+        _activate_async(context, nm_dev)
```

Comment:
I agree with this approach as it makes sense to reapply only if the device is activated. But why wasn't this failing?

Reply:
It will fail and we faill back to normal activation in the `_modify_callback()` to `_activate_async(context, src_object)`